### DISCUSSION TITLE
Setting relationship join condition on nested attributes

### DISCUSSION THREAD

**Discussion by LonelyVikingMichael** [sqlalchemy/sqlalchemy#9896] [Usage Questions] [answered]
Hi, I have 3 tables along with an association table `UserDepartmentRole` as shown below:
```python
class Role(DeclarativeBase):
    __tablename__ = "role"

    name: Mapped[str] = mapped_column(String(255), nullable=False)


class UserDepartmentRole(DeclarativeBase):
    __tablename__ = "user_department_role"

    user_id: Mapped[UUID] = mapped_column(Uuid(), ForeignKey("user.id"), nullable=False)
    department_id: Mapped[UUID] = mapped_column(Uuid(), ForeignKey("department.id"), nullable=False)
    role_id: Mapped[UUID] = mapped_column(Uuid(), ForeignKey("role.id"), nullable=False)

    role: Mapped[Role] = relationship(lazy="joined", uselist=False)
    role_name = association_proxy("role", "name")


class User(DeclarativeBase):
    __tablename__ = "user"

    name: Mapped[str] = mapped_column(String(255), nullable=False)

    department_roles: Mapped[list[UserDepartmentRole]] = relationship(lazy="joined", viewonly=True)


class Department(DeclarativeBase):
    __tablename__ = "department"

    name: Mapped[str] = mapped_column(String(255), nullable=False)

    verifiers: Mapped[list[User]] = relationship(
        lazy="selectin",
        secondary="user_department_role",
        primaryjoin="and_(Department.id == UserDepartmentRole.department_id, UserDepartmentRole.role_name == 'verifier')",
    )
```
I am having trouble setting up the `Department.verifiers` relationship. Here's the error output
```
sqlalchemy.exc.ArgumentError: Could not locate any simple equality expressions involving locally mapped foreign key 
columns for primary join condition 'department.id = user_department_role.department_id AND (EXISTS (SELECT 1 FROM 
role, user_department_role WHERE role.id = user_department_role.role_id AND role.name = :name_1))' on relationship 
Department.users.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are 
annotated in the join condition with the foreign() annotation.
```
I have tried setting `UserDepartmentRole.role_name` as a `hybrid_property` like this, but no luck here either:
```python
    @hybrid_property
    def role_name(self) -> str:
        return self.role.name

    @role_name.inplace.expression
    @classmethod
    def _role_name_expression(cls) -> "SQLColumnExpression[str]":
        return Role.name
```
How should I go about this?

**Comment by zzzeek** [comment, marked as answer]
choice 1:  verifiers would refer to UserDepartmentRole and optionally use [association_proxy](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#module-sqlalchemy.ext.associationproxy) to make access to Department.user more succinct

choice 2: if "secondary" is used, you need to set up a join condition that uses [both primaryjoin and secondaryjoin](https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#composite-secondary-joins) as appropriate between the "local" and "remote" sides


important note:  this part

```
UserDepartmentRole.role_name == 'verifier'
```

that appears to be an assocaition proxy already.  I would not use that in primaryjoin as you can see it is creating an enormously complicated join condition.  please keep these conditions as simple as possible.  also, always use viewonly=True on custom relationships espectially where "secondary" is a table that's mapped (see [important warning at the bottom of this section](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object)


```py
   verifiers: Mapped[list[User]] = relationship(
        lazy="selectin",
        secondary="user_department_role",
        primaryjoin="Department.id == user_department_role.c.department_id",
        secondaryjoin="and_(user_department_role.c.user_id == User.id, user_department_role.c.role_name == 'verifier')",
        viewonly=True
    )

```

**Reply by LonelyVikingMichael** [reply]
Option 1 is the way to go it seems. I was hoping to do something along the lines of option 2 _and_ using that relationship for persistence a.k.a. `viewonly=False`, but it looks like that isn't worth the effort.

Thanks for your insight, Mike.
